File: api/jobs.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { getBackendUrl } from './utils/backendUrl';
import { forwardResponse } from './utils/forwardResponse';
import { getInternalKeyHeader } from './utils/internalKey';
import { PROXY_REJECTION, resolveProxyPath } from './utils/proxyPath';

/**
 * Opaque keyset-pagination token minted by the backend's `GET /api/jobs`.
 *
 * `forwardResponse` copies status + body ONLY, so a response header dies at this
 * proxy unless it is explicitly re-emitted. That failure is silent in exactly the
 * way this pagination work exists to prevent: the array forwards fine, the SPA
 * sees no next-page token, and the walk stops after page 1 looking like a
 * legitimate end-of-results.
 *
 * `GET /api/jobs/search` deliberately does NOT use a header — it returns
 * `{jobs, nextCursor, meta}` — precisely so its cursor cannot be lost this way.
 */
const NEXT_CURSOR_HEADER = 'x-next-cursor';

/**
 * Repeatable query parameters on `GET /api/jobs/search` (multi-select filters).
 *
 * They must be `append`ed rather than `set`, and Vercel hands a repeated param
 * back as `string[]`. Collapsing that array to one value — which `String(x)` on
 * an array quietly does, joining with commas — would turn two selected
 * categories into one bogus slug named `a,b` that matches nothing. Commas are
 * also why these cannot be a comma-joined scalar in the first place: canonical
 * location names ("Austin, TX, US") and free-text keywords contain them.
 */
const REPEATABLE_PARAMS = ['category', 'level', 'company', 'location', 'include', 'exclude'] as const;

export default async function handler(req: VercelRequest, res: VercelResponse) {
  const { path, status, company, companies, limit, offset, category, level, since, cursor } =
    req.query;

  // Sub-path routing (vercel.json rewrites /api/jobs/:path -> ?path=...).
  // Only the facets catalog and the Recent page's search endpoint are exposed;
  // the internal enrichment routes must never be reachable through this public
  // proxy.
  //
  // The allowlist was always right here — a single literal comparison can't be
  // traversed. What it lacked was NORMALIZATION: `?path=facets/` and the array
  // form `?path=facets&path=` both 404'd a legitimate caller. Sharing
  // `resolveProxyPath` with the other seven proxies fixes that and keeps one
  // implementation of the control instead of eight.
  //
  // `:source/:job` is the single-posting detail route `GET
  // /api/jobs/{source_id}/{job_id}`. It stayed off the allowlist until something
  // actually needed it — the WebMCP `get_job` tool now does, so this is the
  // deliberate decision the old comment asked for, not a cleanup. Two dynamic
  // segments matched by the `:param` form `resolveProxyPath` already supports
  // (each `:name` matches exactly one already-canonicalized segment); a
  // traversal like `/api/jobs/../admin` is rejected in canonicalization before
  // it can reach this pattern. The GET-only detail read injects the internal key
  // like every other sub-path and 404s cleanly when no row matches.
  const sub = resolveProxyPath(path, ['', 'facets', 'search', ':source/:job']);
  if (sub === null) {
    res.status(PROXY_REJECTION.status).json(PROXY_REJECTION.body);
    return;
  }

  const params = new URLSearchParams();
  if (status) params.set('status', String(status));
  if (companies) params.set('companies', String(companies));
  if (limit) params.set('limit', String(limit));
  if (offset) params.set('offset', String(offset));
  // Keyset pagination (backend contract: see src/backend/api/routers/jobs.py).
  // `since` bounds the result set by first_seen_at; `cursor` is the opaque token
  // from the previous page's X-Next-Cursor header (or `nextCursor` body field on
  // the search endpoint).
  //
  // Forwarded on PRESENCE, not truthiness, unlike the filters below. `?cursor=`
  // with an empty value is a client bug the backend answers with a 422; a
  // truthiness check would drop it here and hand back page 1 with a 200 instead —
  // converting a loud error into a silent restart of the caller's paging loop.
  if (since !== undefined) params.set('since', String(since));
  if (cursor !== undefined) params.set('cursor', String(cursor));

  if (sub === 'search') {
    // The search endpoint takes every facet filter as a REPEATABLE param, so it
    // cannot share the scalar handling below. Presence-based for the same reason
    // as since/cursor: `?include=` is a caller bug the backend reports as a 422,
    // and silently dropping it would hand back an unfiltered result set the
    // caller believes was filtered.
    for (const name of REPEATABLE_PARAMS) {
      const value = req.query[name];
      if (value === undefined) continue;
      for (const item of Array.isArray(value) ? value : [value]) {
        params.append(name, String(item));
      }
    }
  } else {
    if (company) params.set('company', String(company));
    // Enrichment facet filters, single-valued on the legacy list endpoint.
    // Truthy-checked, unlike the search params above: on this endpoint an empty
    // filter genuinely means "no filter", so dropping it is the correct reading.
    if (category) params.set('category', String(category));
    if (level) params.set('level', String(level));
  }

  const backendUrl = getBackendUrl(req);
  const queryString = params.size ? `?${params}` : '';
  const url = `${backendUrl}/api/jobs${sub ? `/${sub}` : ''}${queryString}`;

  try {
    const response = await fetch(url, {
      headers: getInternalKeyHeader(),
      // Never follow a redirect — Node's fetch preserves headers (including the
      // injected X-Internal-Key) across a same-origin 3xx. See api/jobs-qa.ts.
      redirect: 'manual',
    });
    // Must be set BEFORE forwardResponse — that helper ends the response.
    // Absent on the last page and on the legacy (no since/cursor) path, which is
    // the contract's "end of results" signal, so only copy it when present.
    const nextCursor = response.headers.get(NEXT_CURSOR_HEADER);
    if (nextCursor) res.setHeader('X-Next-Cursor', nextCursor);
    // Edge-cache ONLY the facets catalog. It is effectively immutable — the
    // enrichment taxonomy changes only on a migration+deploy — so a full-day
    // edge TTL with a week of stale-while-revalidate removes the ~0.7 s
    // Vercel->Railway hop from every visitor after the first. `Cache-Control`
    // stays `max-age=0` so browsers always revalidate to the edge and a purge
    // is instantly visible; only `Vercel-CDN-Cache-Control` pins the edge copy.
    // Set BEFORE forwardResponse (it ends the response), and gated so `search`
    // (append-heavy, the slow query) and the legacy company list stay uncached.
    // Also gated on `response.ok`: forwardResponse preserves the upstream status,
    // so an error (a 4xx/5xx, or a `redirect:'manual'` 3xx) must not be edge-cached
    // for a day under the facets key.
    if (sub === 'facets' && response.ok) {
      res.setHeader('Cache-Control', 'public, max-age=0, must-revalidate');
      res.setHeader(
        'Vercel-CDN-Cache-Control',
        'public, s-maxage=86400, stale-while-revalidate=604800',
      );
    }
    await forwardResponse(response, res);
  } catch (error) {
    // Logged, never returned: Node's fetch errors carry the internal backend
    // hostname and port, which a public 500 must not expose. Same split as every
    // other proxy in this directory (api/locations.ts, api/admin.ts, …) — this
    // was the one that swallowed the reason entirely, leaving an opaque 500 with
    // nothing on either side of the wire to debug from.
    console.error('[api/jobs] Upstream fetch failed:', error);
    res.status(500).json({ error: 'Failed to fetch from backend' });
  }
}

File: src/backend/api/routers/jobs.py
"""Jobs API endpoints - GET /api/jobs, GET /api/jobs/{source_id}/{id}."""

import re
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Path, Query, Response
from psycopg2.extensions import connection as Connection

from ..dependencies import get_db
from ..models import (
    COMPANY_PATTERN,
    ENABLED_COMPANY_ID_PATTERN,
    FacetOption,
    JobFacetsResponse,
    JobListingResponse,
)
from ..pagination import (
    MAX_CURSOR_LENGTH,
    MAX_TIMESTAMP_LENGTH,
    InvalidCursorError,
    JobCursor,
    decode_job_cursor,
    encode_job_cursor,
    parse_utc_timestamp,
)
from ..services.database import get_jobs, get_job_by_id
from ..services.enrichment_monitor import get_facets

router = APIRouter()

# Response header carrying the opaque cursor for the NEXT page.
#
# Deliberately a header and not a body field: the response is a bare JSON array
# and has been since the endpoint shipped. Wrapping it in
# ``{"items": [...], "nextCursor": ...}`` would break every existing consumer
# (the frontend transformer, the admin QA table, any direct API user) in one go,
# for a field only the paging caller needs. The header keeps the body contract
# untouched and makes cursor support purely additive.
#
# Delivery path, all THREE hops of which are handled in this PR. Any one of them
# missing means the header silently never reaches the client:
#   * ``api/jobs.ts`` (the Vercel proxy) explicitly re-emits it — ``forwardResponse``
#     copies status + body only, so without that line the header dies at the proxy.
#   * ``CORSMiddleware`` in ``api/main.py`` lists it in ``expose_headers`` — a
#     browser cannot read a non-safelisted response header cross-origin otherwise.
#   * ``vercel.json``'s ``/api/(.*)`` header block adds
#     ``Access-Control-Expose-Headers`` for cross-origin callers of the PROXY,
#     which never touch the FastAPI CORS middleware at all.
NEXT_CURSOR_HEADER = "X-Next-Cursor"

# Max IDs accepted in `?companies=a,b,c` to bound query size and prevent
# unbounded `IN`-list scans. Recent Jobs fans out across all backend-scraper
# companies — 102 today (Greenhouse + Ashby + Lever + Gem + Eightfold +
# Google/Apple/Microsoft). 150 keeps the original ~50% headroom posture
# (cap was 100 against 49 companies when added). The frontend chunks
# requests at 50 IDs/call so this server-side cap is a defense-in-depth
# bound, not the hot path.
_MAX_COMPANIES_PER_REQUEST = 150
_COMPANY_ID_RE = re.compile(ENABLED_COMPANY_ID_PATTERN)


@router.get(
    "",
    response_model=list[JobListingResponse],
    # Declared so /docs and the generated OpenAPI schema show the end-of-walk
    # signal. A paging client that does not know the header exists cannot page at
    # all, and the header is the ONLY thing that tells it when to stop — leaving it
    # undocumented would make the contract discoverable only by reading this file.
    responses={
        200: {
            "headers": {
                NEXT_CURSOR_HEADER: {
                    "description": (
                        "Opaque token for the next page. Present iff this page came "
                        "back full (len == limit) and keyset paging was requested "
                        "(`since` and/or `cursor`). ABSENT means end of results — "
                        "it is the only termination signal. Echo it back verbatim "
                        "as `?cursor=`."
                    ),
                    "schema": {"type": "string"},
                }
            }
        }
    },
)
def list_jobs(
    response: Response,
    conn: Connection = Depends(get_db),
    company: str | None = Query(default=None, pattern=COMPANY_PATTERN),
    companies: str | None = Query(
        default=None,
        description=(
            "Comma-separated list of company IDs. Mutually exclusive with "
            "`company`. Max 150 IDs."
        ),
        max_length=4096,
    ),
    status: str | None = Query(default=None, pattern=r"^(OPEN|CLOSED)$"),
    category: str | None = Query(
        default=None, pattern=r"^[a-z_]{1,40}$",
        description="Enrichment category slug (e.g. software_engineering).",
    ),
    level: str | None = Query(
        default=None, pattern=r"^[a-z_]{1,20}$",
        description="Enrichment level slug; 'entry' also matches new_grad (new_grad⊂entry).",
    ),
    # Cap accommodates the Recent Jobs page's batched fetch across all
    # backend-scraper companies (~16k+ OPEN rows at the time of writing) in
    # one round trip. The per-company default remains 5000.
    limit: int = Query(default=5000, ge=1, le=50000),
    offset: int = Query(default=0, ge=0),
    since: str | None = Query(
        default=None,
        max_length=MAX_TIMESTAMP_LENGTH,
        description=(
            "Recency lower bound, INCLUSIVE: only jobs with "
            "`first_seen_at >= since`. ISO-8601 with a UTC offset "
            "(e.g. 2026-05-07T00:00:00Z). No server default — omitting it means "
            "no lower bound, i.e. today's behaviour. Presence switches the "
            "endpoint into keyset-paging mode."
        ),
    ),
    cursor: str | None = Query(
        default=None,
        max_length=MAX_CURSOR_LENGTH,
        description=(
            "Opaque page token echoed back from a previous response's "
            "`X-Next-Cursor` header. Treat it as a blob. Presence switches the "
            "endpoint into keyset-paging mode."
        ),
    ),
) -> list[JobListingResponse]:
    """List jobs with optional filtering by company and status.

    Accepts either a single ``company`` or a comma-separated ``companies``
    list (for batched per-company fetches from the Recent Jobs page).
    Passing both is a 400.

    KEYSET PAGINATION (``since`` / ``cursor``)
    ------------------------------------------
    Passing either parameter switches this endpoint from ``ORDER BY
    last_seen_at DESC`` to the bounded keyset ordering ``(first_seen_at DESC,
    source_id DESC, id DESC)``. Passing NEITHER is byte-identical to the
    pre-keyset behaviour — same SQL, same ordering, same response, no new header
    — so every existing caller is untouched.

    Response shape is unchanged either way: a **bare JSON array**. The token for
    the next page rides in the ``X-Next-Cursor`` response header:

    * **Present** when the page came back full (``len(page) == limit``), meaning
      more rows may exist. Re-issue the same request with ``cursor`` set to that
      value.
    * **Absent** when the page came back short — that is the end of the walk, and
      the only end signal. A trailing exactly-full page therefore costs one extra
      round trip that returns an empty array; that is the standard keyset
      trade-off, and it is preferred over guessing (a ``LIMIT limit + 1`` probe
      would make the endpoint over-fetch on every single page).
    * Never emitted on the legacy (no ``since``, no ``cursor``) path — a cursor
      minted from a ``last_seen_at``-ordered page would point at the wrong
      boundary in the ``first_seen_at`` ordering and silently skip rows.

    To page the full corpus with no recency bound, pass a floor such as
    ``since=1970-01-01T00:00:00Z``; there is no separate "enable paging" switch.

    Both parameters are validated **fail-loud**: anything malformed is a 422 with
    a specific reason, never a silently-ignored parameter. A dropped cursor would
    restart the walk at page 1 with no signal to the client.

    ``offset`` is **rejected (422) in keyset mode**. The two are different answers
    to the same question — "where does this page start?" — and applying both means
    the cursor seeks to the boundary and ``OFFSET n`` then throws away the first
    ``n`` rows *after* it. That is silent row loss with a 200, the exact failure
    keyset paging exists to remove. ``offset=0`` is fine (it is the default and a
    no-op); the legacy path is untouched and still supports ``offset`` normally.

    A cursor is only meaningful **under the same filter set it was minted with**.
    Changing ``companies`` / ``status`` / ``category`` / ``level`` / ``since``
    partway through a walk is not an error and does not corrupt anything — the
    cursor names a position in the sort order, which is filter-independent — but
    the resulting pages are relative to the NEW filters, so the walk is no longer
    a complete enumeration of either set. Treat a filter change as starting a new
    walk and drop the cursor.

    ``status`` composes with cursors normally (the predicates simply AND
    together). Note that ``idx_job_listings_open_first_seen_keyset`` is partial on
    ``status = 'OPEN'``, so **any request that does not filter to OPEN — including
    one that omits ``status`` entirely, not just ``status=CLOSED`` — falls off it**
    and sorts instead of seeking. Still correct, just unindexed for the ordering.
    Accepted deliberately: every real caller of the paged path filters to OPEN, and
    a full index over both statuses would carry the bloat cost this table's whole
    2026 index history is about.
    """
    parsed_cursor: JobCursor | None = None
    if cursor is not None:
        try:
            parsed_cursor = decode_job_cursor(cursor)
        except InvalidCursorError as exc:
            raise HTTPException(status_code=422, detail=f"Invalid 'cursor': {exc}")

    parsed_since: datetime | None = None
    if since is not None:
        try:
            parsed_since = parse_utc_timestamp(since, field="'since'")
        except ValueError as exc:
            raise HTTPException(status_code=422, detail=f"Invalid 'since': {exc}")

    # See the docstring: offset and a keyset cursor are competing definitions of
    # "where this page starts", and get_jobs applies BOTH (the LIMIT/OFFSET tail is
    # shared by either ordering). The combination therefore skips `offset` rows past
    # the cursor boundary and returns a 200 — silent loss. Reject instead of
    # silently ignoring `offset`, per this router's fail-loud posture.
    if offset and (parsed_cursor is not None or parsed_since is not None):
        raise HTTPException(
            status_code=422,
            detail=(
                "'offset' is incompatible with keyset paging ('cursor'/'since') — "
                "the cursor already determines where the page starts, and applying "
                "both would silently skip rows. Follow 'X-Next-Cursor' instead."
            ),
        )

    company_list: list[str] | None = None
    if companies is not None:
        if company is not None:
            raise HTTPException(
                status_code=400,
                detail="Use either 'company' or 'companies', not both.",
            )
        # Reject empty / whitespace-only values rather than silently treating
        # them as "no filter" — that would be surprising behavior on a typo.
        raw_ids = [c.strip() for c in companies.split(",")]
        if not raw_ids or any(not c for c in raw_ids):
            raise HTTPException(
                status_code=400,
                detail="'companies' must be a non-empty comma-separated list.",
            )
        if len(raw_ids) > _MAX_COMPANIES_PER_REQUEST:
            raise HTTPException(
                status_code=400,
                detail=f"'companies' accepts at most {_MAX_COMPANIES_PER_REQUEST} IDs.",
            )
        for cid in raw_ids:
            if not _COMPANY_ID_RE.match(cid):
                raise HTTPException(
                    status_code=400,
                    detail=f"Invalid company id in 'companies': {cid!r}",
                )
        company_list = raw_ids

    jobs = get_jobs(
        conn,
        company=company,
        companies=company_list,
        status=status,
        limit=limit,
        offset=offset,
        category=category,
        level=level,
        since=parsed_since,
        cursor=parsed_cursor,
    )

    # Mint the next-page token only in keyset mode, and only off a FULL page —
    # see the docstring. ``jobs[-1]`` is the last row of the current page in the
    # keyset ordering, so it is exactly the boundary the next request resumes
    # after.
    keyset_mode = parsed_since is not None or parsed_cursor is not None
    if keyset_mode and len(jobs) == limit:
        tail = jobs[-1]
        response.headers[NEXT_CURSOR_HEADER] = encode_job_cursor(
            tail["first_seen_at"], tail["source_id"], tail["id"]
        )

    return [JobListingResponse(**job) for job in jobs]


@router.get("/facets", response_model=JobFacetsResponse)
def list_facets(conn: Connection = Depends(get_db)) -> JobFacetsResponse:
    """Dropdown catalog for the enrichment facets, straight from the seeded
    job_categories / job_levels dimensions — labels, ordering and the
    new_grad->entry parent all stay data-driven, so a taxonomy change ships as
    a migration without a frontend redeploy. Declared above the parametrized
    detail route by convention (different segment count, so no actual overlap).
    """
    data = get_facets(conn)
    return JobFacetsResponse(
        categories=[FacetOption(**row) for row in data["categories"]],
        levels=[FacetOption(**row) for row in data["levels"]],
    )


@router.get("/{source_id}/{job_id}", response_model=JobListingResponse)
def get_job(
    source_id: str = Path(max_length=100),
    job_id: str = Path(max_length=200),
    conn: Connection = Depends(get_db),
) -> JobListingResponse:
    """Get a single job by composite ``(source_id, id)`` key.

    Returns 404 if no row matches the composite key.
    """
    job = get_job_by_id(conn, source_id, job_id)
    if job is None:
        raise HTTPException(status_code=404, detail="Job not found")
    return JobListingResponse(**job)
